pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go

```diff
@@ -275,52 +280,281 @@ func getPVCDeploymentsNamespace(
-	err = patchController(controllerObj, ns)
+	err = patchController(j.controllerObj, j.ns)
+	if err != nil {
+		statusObj.Message = "failed to patch target depoyment"
+		statusObj.Reason = strings.Replace(err.Error(), ":", "", -1)
+		j.utaskObj, uerr = updateUpgradeDetailedStatus(j.utaskObj, statusObj, openebsNamespace)
+		if uerr != nil && isENVPresent {
+			return uerr
+		}
+		return err
+	}
+	pvLabel := "openebs.io/persistent-volume=" + pvName
+	serviceLabel := "openebs.io/controller-service=jiva-controller-svc," + pvLabel
+
+	err = patchService(serviceLabel, j.ns)
 	if err != nil {
+		statusObj.Message = "failed to patch target service"
+		statusObj.Reason = strings.Replace(err.Error(), ":", "", -1)
+		j.utaskObj, uerr = updateUpgradeDetailedStatus(j.utaskObj, statusObj, openebsNamespace)
+		if uerr != nil && isENVPresent {
+			return uerr
+		}
 		return err
 	}
 
-	err = patchService(serviceLabel, ns)
+	statusObj.Phase = utask.StepCompleted
+	statusObj.Message = "Target upgrade was successful"
+	statusObj.Reason = ""
+	j.utaskObj, uerr = updateUpgradeDetailedStatus(j.utaskObj, statusObj, openebsNamespace)
+	if uerr != nil && isENVPresent {
+		return uerr
+	}
+	return nil
+}
+
+func (j *jivaVolumeOptions) verify(controllerLabel, openebsNamespace string) error {
+	var err, uerr error
+	statusObj := utask.UpgradeDetailedStatuses{Step: utask.Verify}
+	statusObj.Phase = utask.StepWaiting
+	j.utaskObj, uerr = updateUpgradeDetailedStatus(j.utaskObj, statusObj, openebsNamespace)
+	if uerr != nil && isENVPresent {
+		return uerr
+	}
+
+	statusObj.Phase = utask.StepErrored
+	// Verify synced replicas
+	err = validateSync(controllerLabel, j.ns)
 	if err != nil {
+		statusObj.Message = "failed to verify synced replicas"
+		statusObj.Reason = strings.Replace(err.Error(), ":", "", -1)
+		j.utaskObj, uerr = updateUpgradeDetailedStatus(j.utaskObj, statusObj, openebsNamespace)
+		if uerr != nil && isENVPresent {
+			return uerr
+		}
 		return err
 	}
 
-	fmt.Println("Upgrade Successful for", pvName)
+	statusObj.Phase = utask.StepCompleted
+	statusObj.Message = "Replica sync was successful"
+	statusObj.Reason = ""
+	j.utaskObj, uerr = updateUpgradeDetailedStatus(j.utaskObj, statusObj, openebsNamespace)
+	if uerr != nil && isENVPresent {
+		return uerr
+	}
 	return nil
 }
+
+func jivaUpgrade(pvName, openebsNamespace string) (*utask.UpgradeTask, error) {
+
+	var (
+		pvLabel         = "openebs.io/persistent-volume=" + pvName
+		controllerLabel = "openebs.io/controller=jiva-controller," + pvLabel
+		err             error
+	)
+
+	options := &jivaVolumeOptions{}
+
+	// PreUpgrade
+	err = options.preupgrade(pvName, openebsNamespace)
+	if err != nil {
+		return options.utaskObj, err
+	}
+
+	// ReplicaUpgrade
+	err = options.replicaUpgrade(openebsNamespace)
+	if err != nil {
+		return options.utaskObj, err
+	}
+
+	// TargetUpgrade
+	err = options.targetUpgrade(pvName, openebsNamespace)
+	if err != nil {
+		return options.utaskObj, err
+	}
+
+	// Verify
```

Comment:
It is possible that on some setups pod/exec will be disabled. In this case, we should patch the status with a warning message and provide information on how to verify. Alternatively, we can probably use the m-apiserver api to list the jiva volume and verify the status. Also, can be taken in a different PR.

Reply:
Added a check for the cases in which `exec` into the pods is `Forbidden`. This required removing wrapper error from some functions as wrapping the error was removing the reason of the error.